pkgs/by-name/ld/ld64/package.nix

```diff
@@ -0,0 +1,184 @@
+    name = "xnu-src";
+    owner = "apple-oss-distributions";
+    repo = "xnu";
+    rev = "xnu-6153.11.26";
+    hash = "sha256-dcnGcp7bIjQxeAn5pXt+mHSYEXb2Ad9Smhd/WUG4kb4=";
+  };
+  # Avoid pulling in all of Swift just to build libdispatch
+  libdispatch = swiftPackages.Dispatch.override { useSwift = false; };
+in
+stdenv.mkDerivation (finalAttrs: {
+  pname = "ld64";
+  version = "951.9";
+
+  outputs = [
+    "out"
+    "dev"
+    "lib"
+  ];
+
+  src = fetchFromGitHub {
+    owner = "apple-oss-distributions";
+    repo = "ld64";
+    rev = "ld64-${finalAttrs.version}";
+    hash = "sha256-hLkfqgBwVPlO4gfriYOawTO5E1zSD63ZcNetm1E5I70";
+  };
+
+  sourceRoot = "source";
+
+  postUnpack = ''
+    unpackFile '${xnu}'
+  '';
+
+  patches = [
+    # Use std::atomic for atomics.
+    ./0004-Use-std-atomics-and-std-mutex.patch
+    # ld64 assumes the default libLTO.dylib can be found relative to its bindir, which is
+    # not the case in nixpkgs. Override it to default to `stdenv.cc`’s libLTO.dylib.
+    ./0005-Support-LTO-in-nixpkgs.patch
+    # Add implementation of missing function required for code directory support.
+    ./0006-Add-libcd_is_blob_a_linker_signature-implementation.patch
+    # Add OpenSSL implementation of CoreCrypto digest functions
+    ./0007-Add-OpenSSL-based-CoreCrypto-digest-functions.patch
+    # ld64 will search `/usr/lib`, `/Library/Frameworks`, etc by default. Disable that.
+    ./0008-Disable-searching-in-standard-library-locations.patch
+  ];
+
+  postPatch = ''
+    substitute ${./meson.build} meson.build \
+      --subst-var-by targetPrefix '${targetPrefix}' \
+      --subst-var version
+
+    # Copy headers for certain private APIs
+    mkdir -p include
+    substitute ${crashreporter_h} include/CrashReporterClient.h \
+      --replace-fail 'USE(APPLE_INTERNAL_SDK)' '0'
+
+    # Copy from the source so the headers can be used on Linux and x86_64-darwin
+    mkdir -p include/System
+    for dir in arm i386 machine; do
+      cp -r ../xnu-src/osfmk/$dir include/System/$dir
+    done
+    mkdir -p include/sys
+    cp ../xnu-src/bsd/sys/commpage.h include/sys
+
+    # Match the version format used by upstream.
+    sed -i src/ld/Options.cpp \
+      -e '1iconst char ld_classicVersionString[] = "@(#)PROGRAM:ld PROJECT:ld64-${finalAttrs.version}\\n";'
+
+    # Instead of messing around with trying to extract and run the script from the Xcode project,
+    # just use our own Python script to generate `compile_stubs.h`
+    cp ${./gen_compile_stubs.py} gen_compile_stubs.py
+
+    # Enable LTO support using LLVM’s libLTO.dylib by default.
+    substituteInPlace src/ld/InputFiles.cpp \
+      --subst-var-by libllvm '${lib.getLib llvm}'
+    substituteInPlace src/ld/parsers/lto_file.cpp \
+      --subst-var-by libllvm '${lib.getLib llvm}'
+
+    # Use portable includes
+    substituteInPlace src/ld/code-sign-blobs/endian.h \
+      --replace-fail '#include <machine/endian.h>' '#include <sys/types.h>'
```

Comment:
The comments should be self-explanatory. It’s used on Darwin with an eye towards Linux (which is not complete because I wanted to get something done for Darwin because that’s what is actually important).

Reply:
i feel like i am missing the core concept of why this needs to work on linux.

Reply:
Because people want to cross-compile to Darwin from Linux.

Reply:
While I would like Linux→Darwin cross support, maybe it makes sense to make these changes when the rest of the pieces for that are in place and we can assess the total cost‐benefit, to keep the complexity/review burden down for now? What are the disadvantages of depending on the XNU headers currently?

Reply:
It’s not clear to me how vendoring a few headers increases the complexity of review, but the 10.12 SDK xnu headers are too old.

Reply:
I was thinking more the OpenSSL and `std::atomics` stuff, probably should have commented on those instead.

Reply:
The `std::atomics` change was done for Darwin, though the solution was chosen with Linux in mind. ld64 uses `os/lock_private.h`, which is not available in the 11.0 SDK in nixpkgs. I could vendor it, but I prefer to replace private APIs with non-private ones when I encounter them.

The digest change (to use OpenSSL) was done for Linux support before I realized the scope and stopped, but it also removes use of a several private APIs (corecrypto, `CommonCrypto/CommonDigestSPI.h`). Since ld64 just needs to calculate digest, it write shims to replace usage of those private APIs with OpenSSL.¹

corecrypto in particular is extremely non-free. Apple makes the source [available][2] under a “look but don’t use” license for verifying the implementation. The headers in xnu also have this license.

---

[1]: I opted to use shims instead of patching the usage directly to keep the size of the diff down. cctools-port also [uses][3] this approach, though they support using corecrypto if it’s available. I don’t because I want to use the same implementation on all platforms (but also licensing).

[2]: https://developer.apple.com/security/
[3]: https://github.com/tpoechtrager/cctools-port/tree/f28fb5e9c31efd3d0552afcce2d2c03cae25c1ca/cctools/ld64/src/3rd/compat_corecrypto

Reply:
@emilazy I expanded the comments for both of these. Let me know if that helps address your concerns.